File: src/main/java/gargoyle/rpycg/ui/model/DisplayItem.java
package gargoyle.rpycg.ui.model;

import gargoyle.rpycg.model.ModelType;
import gargoyle.rpycg.model.VarType;
import javafx.beans.property.Property;
import javafx.beans.property.ReadOnlyProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.scene.control.TreeItem;

import java.text.MessageFormat;
import java.util.Objects;

public final class DisplayItem {
    private final Property<String> label;
    private final ReadOnlyProperty<ModelType> modelType;
    private final Property<String> name;
    private final Property<VarType> type;
    private final Property<String> value;

    private DisplayItem(ModelType modelType,
                        String label, String name, String value,
                        VarType type) {
        this.modelType = new SimpleObjectProperty<>(modelType);
        this.label = new SimpleStringProperty(label);
        this.name = new SimpleStringProperty(name);
        this.value = new SimpleStringProperty(value);
        this.type = new SimpleObjectProperty<>(type);
    }

    public static DisplayItem createRoot() {
        return createMenu("", "");
    }

    public static DisplayItem createMenu(String label, String name) {
        return new DisplayItem(ModelType.MENU, label, name, "", null);
    }

    public static DisplayItem createVariable(VarType type,
                                             String label, String name, String value) {
        return new DisplayItem(ModelType.VARIABLE, label, name, value, type);
    }

    public static TreeItem<DisplayItem> toTreeItem(DisplayItem item, boolean expanded) {
        TreeItem<DisplayItem> treeItem = new DisplayTreeItem(item);
        treeItem.setExpanded(expanded);
        return treeItem;
    }

    public DisplayItem copyOf() {
        return new DisplayItem(getModelType(), getLabel(), getName(), getValue(), getType());
    }

    public ModelType getModelType() {
        return modelType.getValue();
    }

    public String getLabel() {
        return label.getValue();
    }

    public String getName() {
        return name.getValue();
    }

    public void setName(String name) {
        this.name.setValue(name);
    }

    public String getValue() {
        return value.getValue();
    }

    public VarType getType() {
        return type.getValue();
    }

    public void setType(VarType type) {
        this.type.setValue(type);
    }

    public void setValue(String value) {
        this.value.setValue(value);
    }

    public void setLabel(String label) {
        this.label.setValue(label);
    }

    @Override
    public int hashCode() {
        return Objects.hash(modelType, name);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || DisplayItem.class != obj.getClass()) {
            return false;
        }
        DisplayItem item = (DisplayItem) obj;
        return modelType.getValue() == item.modelType.getValue() && name.getValue().equals(item.name.getValue());
    }

    @Override
    public String toString() {
        return MessageFormat.format("{0}[{1}]'{'type={2}, label={3}, name={4}, value={5}'}'",
                DisplayItem.class.getSimpleName(),
                modelType.getValue(), type.getValue(), label.getValue(),
                name.getValue(), value.getValue());
    }

    public Property<String> labelProperty() {
        return label;
    }

    public ReadOnlyProperty<ModelType> modelTypeProperty() {
        return modelType;
    }

    public Property<String> nameProperty() {
        return name;
    }

    public Property<VarType> typeProperty() {
        return type;
    }

    public Property<String> valueProperty() {
        return value;
    }

    private static final class DisplayTreeItem extends TreeItem<DisplayItem> {
        private DisplayTreeItem(DisplayItem item) {
            super(item);
        }

        @Override
        public String toString() {
            return "TreeItem: " + getValue();
        }
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/ui/model/DisplayItem.kt
package gargoyle.rpycg.ui.model

import gargoyle.rpycg.model.ModelType
import gargoyle.rpycg.model.VarType
import javafx.beans.property.Property
import javafx.beans.property.ReadOnlyProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.scene.control.TreeItem

@Suppress("MemberVisibilityCanBePrivate")
class DisplayItem private constructor(modelType: ModelType, label: String, name: String, value: String, type: VarType) {

    val label: Property<String>
    val modelType: ReadOnlyProperty<ModelType>
    val name: Property<String>
    val type: Property<VarType>
    val value: Property<String>

    fun copyOf(): DisplayItem =
        DisplayItem(modelType.value, label.value, name.value, value.value, type.value)

    fun getLabel(): String =
        label.value

    fun setLabel(label: String) {
        this.label.value = label
    }

    fun getModelType(): ModelType =
        modelType.value

    fun getName(): String =
        name.value

    fun setName(name: String) {
        this.name.value = name
    }

    fun getType(): VarType =
        type.value

    fun setType(type: VarType) {
        this.type.value = type
    }

    fun getValue(): String =
        value.value

    fun setValue(value: String) {
        this.value.value = value
    }

    override fun hashCode(): Int =
        arrayOf(modelType, name).contentHashCode()

    override fun equals(other: Any?): Boolean =
        if (this === other) true else if (null == other || DisplayItem::class.java != other.javaClass) false
        else modelType.value === (other as DisplayItem).modelType.value && name.value == (other as DisplayItem).name.value

    override fun toString(): String =
        "DisplayItem[${modelType.value}]{type=${type.value}, label=${label.value}, name=${name.value}, value=${value.value}}"

    fun labelProperty(): Property<String> =
        label

    fun modelTypeProperty(): ReadOnlyProperty<ModelType> =
        modelType

    fun nameProperty(): Property<String> =
        name

    fun typeProperty(): Property<VarType> =
        type

    fun valueProperty(): Property<String> =
        value

    private class DisplayTreeItem(item: DisplayItem) : TreeItem<DisplayItem>(item) {
        override fun toString(): String =
            "TreeItem: $value"
    }

    companion object {
        fun createMenu(label: String, name: String): DisplayItem =
            DisplayItem(ModelType.MENU, label, name, "", VarType.STR)

        fun createRoot(): DisplayItem = createMenu("", "")

        fun createVariable(type: VarType, label: String, name: String, value: String): DisplayItem =
            DisplayItem(ModelType.VARIABLE, label, name, value, type)

        fun toTreeItem(item: DisplayItem, expanded: Boolean): TreeItem<DisplayItem> {
            val treeItem: TreeItem<DisplayItem> = DisplayTreeItem(item)
            treeItem.isExpanded = expanded
            return treeItem
        }
    }

    init {
        this.modelType = SimpleObjectProperty(modelType)
        this.label = SimpleStringProperty(label)
        this.name = SimpleStringProperty(name)
        this.value = SimpleStringProperty(value)
        this.type = SimpleObjectProperty(type)
    }
}
